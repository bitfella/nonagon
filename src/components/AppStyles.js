import 'typeface-montserrat';
import { normalize } from 'styled-normalize';
import { createGlobalStyle } from 'styled-components/macro';

const theme = {
  colors: {
    a: '#FFFFFF', // white
    b: '#0028B3', // blue
    c: '#FFB948', // yellow
    d: '#F52174', // purple
    e: '#000000', // black
    f: '#CECECE', // grey
    g: '#353535', // dark grey
    h: '#DFDFDF', // light grey
    apple: {
      systemBlue: '#0A84FF',
      systemGreen: '#30D158'
    },
    spotify: {
      green: '#1DB954',
      greenHover: '#1ED760',
      red: '#E10000',
      redHover: '#F8000A'
    }
  },
  fonts: {
    body: `'Montserrat', sans-serif`
  },
  sizes: {
    phi: 1.61803398875,
    appMaxWidth: 768,
    headerHeight: 60
  }
};

const GlobalStyles = createGlobalStyle`
  ${normalize}

  html {
    color: ${theme.colors.e};
    font-family: ${theme.fonts.body};
    font-size: 1em;
    font-weight: 400;
    line-height: ${theme.sizes.phi};
    background-color: ${theme.colors.a};
  }

  *:focus {
    outline: none;
  }
`;

export { GlobalStyles, theme };
