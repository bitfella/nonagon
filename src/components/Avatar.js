import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components/macro';

const AvatarWrapper = styled.picture`
  display: block;
  width: 150px;
  height: 150px;
  margin: 0 auto;

  svg {
    width: 100%;
    height: auto;

    polygon {
      stroke: ${props => props.theme.colors.a};
    }
  }
`;

const Avatar = ({ src, alt }) => (
  <AvatarWrapper>
    <svg
      viewBox="0 0 100 100"
      xmlns="http://www.w3.org/2000/svg"
      aria-label={alt}
    >
      <defs>
        <pattern
          id="img"
          patternUnits="userSpaceOnUse"
          width="100"
          height="100"
        >
          <image href={src} x="6" y="4" width="89" height="89" />
        </pattern>
      </defs>
      <polygon
        fill="url(#img)"
        strokeWidth="2"
        points="50 3.5 79.5 14.5 95.5 42 90 73 66 93.5 34.5 93.5 10 73 5 42 20 15"
      />
    </svg>
  </AvatarWrapper>
);

Avatar.propTypes = {
  src: PropTypes.string.isRequired,
  alt: PropTypes.string
};

export default Avatar;
