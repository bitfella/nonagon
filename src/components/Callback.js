import React from 'react';
import { Redirect } from 'react-router-dom';
import { appAccessToken } from 'config/spotify';
import { getHashKeyValue } from 'utils';
import Error from 'components/Error';
import Layout from 'components/Layout';

const Callback = () => {
  const accessToken = getHashKeyValue('access_token');

  if (accessToken) {
    localStorage.setItem(appAccessToken, accessToken);

    return <Redirect to="/" />;
  }

  return (
    <Layout>
      <div isLayoutBodyWrapper>
        <Error />
      </div>
    </Layout>
  );
};

export default Callback;
