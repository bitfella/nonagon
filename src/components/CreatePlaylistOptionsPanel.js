import React, { Component } from 'react';
import styled from 'styled-components/macro';
import PropTypes from 'prop-types';
import CreatePlaylistOptionsList from 'containers/CreatePlaylistOptionsList';

const PlaylistOptionsPanel = styled.div`
  position: fixed;
  right: 0;
  bottom: 0;
  left: 0;
  z-index: 9;
  max-width: ${props => props.theme.sizes.appMaxWidth}px;
  margin: 0 auto;
  padding: 1rem 1rem 6rem 1rem;
  box-sizing: border-box;
  color: ${props => props.theme.colors.a};
  background-color: ${props => props.theme.colors.g};
  border-top-right-radius: 2rem;
  border-top-left-radius: 2rem;
  transition: transform 0.5s ease-out;
`;

class CreatePlaylistOptionsPanel extends Component {
  static propTypes = {
    isOpen: PropTypes.bool
  };

  static defaultProps = {
    isOpen: false
  };

  constructor(props) {
    super(props);

    this.panelElem = React.createRef();
    this.openPos = '0';
    this.closePos = '100%';
    this.closeThreshold = 70;
    this.state = {
      position: 0,
      touchStart: 0,
      touchEnd: 0
    };
  }

  closePanel = () => {
    if (this.props.isOpen) this.props.setPanelVisibility(false);
  };

  handleClickOutside = event => {
    if (
      this.panelElem.current &&
      !this.panelElem.current.contains(event.target) &&
      !(
        event.target.closest('button') &&
        event.target.closest('button').id === 'createPlaylistOptionsToggler'
      )
    ) {
      this.closePanel();
    }
  };

  handleScroll = () => {
    this.closePanel();
  };

  panelShouldClose = () => {
    const { touchEnd, touchStart } = this.state;

    return touchEnd - touchStart >= this.closeThreshold;
  };

  tapPanel = event => {
    if (event.target === this.panelElem.current) {
      this.setState(() => ({
        touchStart: event.changedTouches[0].screenY
      }));
    }
  };

  dragPanel = event => {
    if (this.props.isOpen) event.preventDefault();

    if (event.target === this.panelElem.current) {
      const touchThumb =
        event.changedTouches[0].screenY - this.state.touchStart;

      this.setState(() => ({
        position: touchThumb < 0 ? this.openPos : `${touchThumb}px`
      }));
    }
  };

  releasePanel = event => {
    if (event.target === this.panelElem.current) {
      this.setState(
        {
          touchEnd: event.changedTouches[0].screenY
        },
        () => {
          if (this.panelShouldClose()) {
            this.setState(() => ({
              position: this.closePos
            }));

            this.closePanel();
          } else {
            this.setState(() => ({
              position: this.openPos
            }));
          }
        }
      );
    }
  };

  componentDidMount() {
    document.addEventListener('scroll', this.handleScroll);
    document.addEventListener('touchstart', this.tapPanel);
    document.addEventListener('touchmove', this.dragPanel, { passive: false });
    document.addEventListener('touchend', this.releasePanel);
    document.addEventListener('mousedown', this.handleClickOutside);
  }

  componentWillUnmount() {
    document.removeEventListener('scroll', this.handleScroll);
    document.removeEventListener('touchstart', this.tapPanel);
    document.removeEventListener('touchmove', this.dragPanel);
    document.removeEventListener('touchend', this.releasePanel);
    document.removeEventListener('mousedown', this.handleClickOutside);
  }

  componentDidUpdate(prevProps) {
    if (prevProps.isOpen !== this.props.isOpen) {
      this.setState(() => ({
        position: this.props.isOpen ? this.openPos : this.closePos
      }));
    }
  }

  render() {
    const { position } = this.state;

    return (
      <PlaylistOptionsPanel
        ref={this.panelElem}
        style={{
          transform: `translateY(${
            this.props.isOpen ? position : this.closePos
          })`
        }}
      >
        <CreatePlaylistOptionsList />
      </PlaylistOptionsPanel>
    );
  }
}

export default CreatePlaylistOptionsPanel;
