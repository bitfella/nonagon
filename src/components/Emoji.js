import React from 'react';
import PropTypes from 'prop-types';

const Emoji = ({ label, symbol }) => (
  <span role="img" aria-label={label}>
    {symbol}
  </span>
);

Emoji.propTypes = {
  label: PropTypes.string.isRequired,
  symbol: PropTypes.string.isRequired
};

export default Emoji;
