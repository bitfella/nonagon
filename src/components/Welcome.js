import React, { Component } from 'react';
import { sample } from 'lodash-es';
import PropTypes from 'prop-types';
import styled, { withTheme } from 'styled-components/macro';
import Avatar from 'components/Avatar';
import { getAlternativeImageSrc } from 'utils';

export const greetingsList = [
  'Howdy',
  'Welcome',
  'Hey',
  'Hail',
  'Ciao',
  'Aloha'
];

const WelcomeWrapper = styled.div`
  padding-bottom: 0.8rem;
  background-color: ${props => props.theme.colors.b};
`;

const Message = styled.h3`
  margin-top: 0.5rem;
  margin-bottom: 0;
  color: ${props => props.theme.colors.a};
  font-size: 1rem;
  font-weight: 400;
  text-align: center;

  span {
    color: ${props => props.theme.colors.c};
    font-weight: 600;
  }
`;

export class Welcome extends Component {
  static propTypes = {
    userDisplayName: PropTypes.string.isRequired,
    userImages: PropTypes.array,
    greeting: PropTypes.string
  };

  render() {
    const { userDisplayName, userImages, greeting } = this.props;
    const salute = greeting ? greeting : sample(greetingsList);

    if (userDisplayName && userImages) {
      return (
        <WelcomeWrapper>
          {userImages && userImages.length >= 1 ? (
            <Avatar src={userImages[0].url} alt={userDisplayName} />
          ) : (
              <Avatar
                src={getAlternativeImageSrc(
                  userDisplayName,
                  300,
                  300,
                  this.props.theme.colors.f,
                  this.props.theme.colors.g,
                  this.props.theme.fonts.body,
                  '5rem'
                )}
                alt={userDisplayName}
              />
            )}
          <Message>
            {salute}, <span>{userDisplayName}</span>!
          </Message>
        </WelcomeWrapper>
      );
    }

    return null;
  }
}

export default withTheme(Welcome);
