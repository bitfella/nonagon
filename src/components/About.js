import React from 'react';
import styled from 'styled-components/macro';
import {
  spotifyWebApiScopes,
  spotifyWebApiScopesGuideURI
} from 'config/spotify';
import Emoji from 'components/Emoji';
import Layout from 'components/Layout';

const AboutWrapper = styled.article`
  color: ${props => props.theme.colors.g};
  padding: 1.5rem;
`;

const AboutSection = styled.section`
  p,
  ul {
    margin: 0.625rem 0;
    font-size: 0.875rem;

    &:first-child {
      margin-top: 0;
    }

    &:last-child {
      margin-bottom: 0;
    }
  }

  a {
    color: ${props => props.theme.colors.c};
    font-weight: 600;
  }
`;

const AboutHeading = styled.h3`
  margin-bottom: 0.625rem;
  color: ${props => props.theme.colors.g};
  font-size: 1rem;
`;

const AboutSubHeading = styled.h4`
  margin-bottom: 0.625rem;
  color: ${props => props.theme.colors.g};
  font-size: 0.875rem;
`;

const HighlightNonagon = styled.span`
  color: ${props => props.theme.colors.d};
  font-weight: 600;
`;

const HighlightSpotify = styled.span`
  color: ${props => props.theme.colors.spotify.green};
  font-weight: 600;
`;

const About = () => (
  <Layout>
    <div isLayoutBodyWrapper>
      <AboutWrapper>
        <AboutSection>
          <p>
            <HighlightNonagon>Nonagon</HighlightNonagon> is a little web mobile
            app that helps you creating{' '}
            <HighlightSpotify>Spotify</HighlightSpotify> playlists based upon
            your most listened songs.
          </p>
          <p>
            In cheap words, it reads your Spotify top tracks and generates
            playlists made of Spotify recommendations based on your personal
            tastes.
          </p>
          <AboutHeading>
            Playlist creation options{' '}
            <Emoji symbol="🎛️" label="control knobs" />
          </AboutHeading>
          <p>
            Nonagon gives you the ability to customize your generated playlist
            adjusting the following options in the control panel:
          </p>
          <AboutSubHeading>Make playlist public</AboutSubHeading>
          <p>
            By default, Nonagon playlists are private. If you wish to create
            public playlists, turn the switch on. Please be aware that you can
            change again this value whenever you wish, in your Spotify app.
          </p>
          <AboutSubHeading>Playlist danceability</AboutSubHeading>
          <p>
            Danceability describes how suitable a track is for dancing based on
            a combination of musical elements including tempo, rhythm stability,
            beat strength, and overall regularity. A value of 0% is least
            danceable and 100% is most danceable.
          </p>
          <AboutSubHeading>Playlist energy</AboutSubHeading>
          <p>
            Energy is a measure from 0% to 100% and represents a perceptual
            measure of intensity and activity. Typically, energetic tracks feel
            fast, loud, and noisy. For example, death metal has high energy,
            while a Bach prelude scores low on the scale. Perceptual features
            contributing to this attribute include dynamic range, perceived
            loudness, timbre, onset rate, and general entropy.
          </p>
          <AboutSubHeading>Playlist valence</AboutSubHeading>
          <p>
            A measure from 0% to 100% describing the musical positiveness
            conveyed by a track. Tracks with high valence sound more positive
            (e.g. happy, cheerful, euphoric), while tracks with low valence
            sound more negative (e.g. sad, depressed, angry).
          </p>
          <AboutHeading>
            Why Nonagon? <Emoji symbol="🤔" label="thinking face" />
          </AboutHeading>
          <p>
            Nonagon was named after one of my favourite records, &ldquo;Nonagon
            Infinity&rdquo; by aussie psych rockers{' '}
            <strong>King Gizzard &amp; the Lizard Wizard</strong>{' '}
            <Emoji symbol="🤘" label="rock on" />
          </p>
          <AboutHeading>
            Boring technical stuff <Emoji symbol="😴" label="sleeping face" />
          </AboutHeading>
          <p>
            Nonagon is a client side web application: it does not have any
            backend and it relies totally on Spotify Web Apis to get data. It
            does not save your personal data wherever, although it does save
            playlist on your Spotify account.
          </p>
          <p>
            Nonagon makes use of the following Spotify authorization scopes:
          </p>
          <ul>
            {spotifyWebApiScopes.map((scope, i) => (
              <li key={i}>{scope}</li>
            ))}
          </ul>
          <p>
            Learn more about Spotify Scopes{' '}
            <a
              href={spotifyWebApiScopesGuideURI}
              rel="noopener noreferrer"
              target="_blank"
              title="Spotify Scopes"
            >
              here
            </a>
            .
          </p>
          <AboutHeading>
            Privacy <Emoji symbol="🔐" label="locked with key" />
          </AboutHeading>
          <p>
            Nonagon does not save any cookie on your browser (apart from Spotify API ones) and does not use
            any kind of tracking platform.
          </p>
          <p>
            Nonagon just saves a Spotify authorization token and some of your
            app preferences on local storage: this is mandatory to make the app
            work. You can easily delete them clearing your browser data.
          </p>
        </AboutSection>
      </AboutWrapper>
    </div>
  </Layout>
);

export default About;
