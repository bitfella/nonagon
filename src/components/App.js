import React from 'react';
import PropTypes from 'prop-types';
import { ConnectedRouter } from 'connected-react-router';
import { ThemeProvider } from 'styled-components/macro';
import { GlobalStyles, theme } from 'components/AppStyles';
import Router from 'components/Router';

const App = ({ history }) => (
  <ConnectedRouter history={history}>
    <GlobalStyles />
    <ThemeProvider theme={theme}>
      <Router />
    </ThemeProvider>
  </ConnectedRouter>
);

App.propTypes = {
  history: PropTypes.object
};

export default App;
