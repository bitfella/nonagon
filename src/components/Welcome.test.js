import React from 'react';
import { shallow } from 'enzyme';
import { Welcome, greetingsList } from './Welcome';
import { theme } from 'components/AppStyles';

const props = {
  greeting: greetingsList[0],
  userId: 'id',
  userImages: [],
  theme
};

describe('Welcome', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<Welcome {...props} />);

    expect(wrapper).toMatchSnapshot();
  });
});
