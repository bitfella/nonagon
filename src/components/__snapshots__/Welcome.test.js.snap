// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Welcome renders correctly 1`] = `""`;
